service/history/workflow/update/update.go

```diff
@@ -566,8 +585,16 @@ func (u *Update) onResponseMsg(
 		if !u.state.Matches(stateSet(stateAccepted | stateProvisionallyCompleted)) {
 			return
 		}
-		u.setState(stateCompleted)
+		beforeCommitState := u.setState(stateCompleted)
 		u.outcome.(*future.FutureImpl[*updatepb.Outcome]).Set(res.GetOutcome(), nil)
+		if beforeCommitState == stateAccepted {
+			// If Update is accepted and completed on the same WFT, then state transitions are:
+			// ProvisionallyAccepted -> ProvisionallyCompleted -> Accepted -> Completed.
+			// To prevent race condition with checks in WaitLifecycleStage, accepted future
+			// must be set after outcome future. Therefore, set accepted future here
+			// if state transition is Accepted -> Completed.
+			u.accepted.(*future.FutureImpl[*failurepb.Failure]).Set(nil, nil)
+		}
 		u.onComplete()
 	})
 	eventStore.OnAfterRollback(func(context.Context) {
```

Comment:
Could you add these two new special cases to the `TestUpdateState` in `update_test.go`?